### PR TITLE
send: run callback in case of congestion #41

### PR DESCRIPTION
**Pull request by jbdemonte**

### CODE REVIEW

**Comment by santigimeno** [comment]
@jbdemonte tests don't pass with this change. Could you update them and / or add some new ones for the new functionality? Also, this could be considered a semver-major change. @bnoordhuis WDYT?

**Comment by jbdemonte** [comment]
no problem, I will, I ran tests, I didn't notice error thrown (I usually use mocha :) )

**Comment by santigimeno** [comment]
@jbdemonte there's no error thrown, but the test doesn't exit gracefully ( I used: `npm test` BTW)

**Comment by jbdemonte** [comment]
I updated the unit tests, let me know if it's ok on your side

**Comment by jbdemonte** [comment]
Is there any problem with my PR?

**Comment by bnoordhuis** [comment]
I can't speak for @santigimeno but I probably won't have time to look at it this side of New Year's Eve.

**Comment by santigimeno** [comment]
@jbdemonte it's on my radar but haven't had the time to look at it yet. I think I'll be able to throughout the week.

**Comment by jbdemonte** [comment]
Thanks guys, have good time for theses

File: lib/unix_dgram.js [hunk 1]
```diff
@@ -89,17 +96,20 @@ Socket.prototype.send = function(buf, offset, length, path, callback) {
...
+      callback(internalError(err, 'congestion'));
     } else {
-      this.emit('error', error);
+      callback();
+    }
+  } else {
+    if (err < 0) {
+      this.emit('error', errnoException(err, 'send'));
+    } else if (err === 1) {
+      this.emit('congestion', buf);
```
Comment (santigimeno) [inline lib/unix_dgram.js]: I think this should be `this.emit('congestion');`

File: test/test-connect-callback.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+process.on('exit', function() {
+  assert.equal(seenCount, sentCount);
+});
+
+try { fs.unlinkSync(SOCKNAME); } catch (e) { /* swallow */ }
+
+var server = unix.createSocket('unix_dgram', function(buf, rinfo) {
+  console.error('server recv', '' + buf, arguments);
+  assert.equal('' + buf, 'PING' + seenCount);
```
Comment (santigimeno) [inline test/test-connect-callback.js]: I would remove all the logs from both files as now the tests are too noisy

File: test/test-connect-callback.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        throw new Error(err);
+      } else {
+        console.log('congestion');
+        client.once('writable', send);
+      }
+    });
+  }
+
+  send();
+
```
Comment (santigimeno) [inline test/test-connect-callback.js]: remove extra line

File: test/test-connect-callback.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  client.on('writable', function() {
+    console.error('writable');
+  });
+
+  /*
+   * Usually /proc/sys/net/unix/max_dgram_qlen is 10 so one that's being processed, 10 in the
+   * recv queue and the ... is dropped
+   */
+
```
Comment (santigimeno) [inline test/test-connect-callback.js]: remove extra line

File: test/test-connect-callback.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  });
+
+  client.on('writable', function() {
+    console.error('writable');
+  });
+
+  /*
+   * Usually /proc/sys/net/unix/max_dgram_qlen is 10 so one that's being processed, 10 in the
+   * recv queue and the ... is dropped
+   */
```
Comment (santigimeno) [inline test/test-connect-callback.js]: Maybe add here that this only applies on linux?

File: lib/unix_dgram.js (hunk 1 above)
Comment (jbdemonte) [inline lib/unix_dgram.js]: Thanks to the buf parameter, you're able to resend it in the listeneer

Comment (santigimeno) [inline lib/unix_dgram.js]: Good idea

File: test/test-connect-callback.js (hunk 2 above)
Comment (santigimeno) [inline test/test-connect-callback.js]: Yes, I know. It was already noisy, but now that we're sending 300 messages I think it's too much.

Comment (jbdemonte) [inline test/test-connect-callback.js]: haha, for sure, I agree :D

File: test/test-connect-callback.js (hunk 5 above)
Comment (jbdemonte) [inline test/test-connect-callback.js]: To be honest, I copied this comment from the test-connect.js file, without having being able to see it in action

**Comment by jbdemonte** [comment]
Sorry for the delay for your others comments, I didn't notice the requested changes
Btw, Happy New Year ;)

**Comment by jbdemonte** [comment]
Do you need something else for this PR?

**Comment by santigimeno** [comment]
@jbdemonte, sorry for the delay, I've been AFK most of the holiday season. I'll try yo look into this tomorrow. Thanks!

File: test/test-connect-callback.js (hunk 5 above)
Comment (santigimeno) [inline test/test-connect-callback.js]: I know, I added it initially but we found out in https://github.com/bnoordhuis/node-unix-dgram/issues/39 tht the behavior in Linux is different from `OS X`, also, the `proc` filesystem does not exist on `OS X`. So i think that the comment should be amended or just removed (and also on `test-connect.js`)

**Comment by santigimeno** [comment]
@jbdemonte can you rebase to current master and squash the PR into a single commit? Thanks!

**Comment by jbdemonte** [comment]
It should be ok

**Comment by bnoordhuis** [comment]
@santigimeno If you +1'd it, by all means go ahead and merge it.

**Comment by santigimeno** [comment]
Landed in https://github.com/bnoordhuis/node-unix-dgram/commit/3d9b871e5a1bf3fdca7cf83988d1ffc6702eda68 and published in `v2.0.0`. Thanks!

**Comment by jbdemonte** [comment]
Great :D
It seems to not be published yet on npm, is it voluntary?

**Comment by santigimeno** [comment]
Sorry, I thought I had. Let me check

**Comment by santigimeno** [comment]
@jbdemonte, it should be good now

**Comment by jbdemonte** [comment]
Great, thanks

**Pull request closed without merging**